### PR TITLE
Liblist

### PR DESCRIPTION
**Pull request by Athinanarof**
## What is the problem/goal being addressed?
ListLib functions implemented

## What is the solution to this problem?
Functions Implementation and testing.

## What are the changes here? How do they solve the problem and what other product impacts do they cause?
*Please include before/after screenshots/gifs if appropriate*
LibList functions Implementation and tested.

## How are you sure this works/how was this tested?
I created the test cases for each function.

## What is the reversion plan if this fails after shipping?
Revert branch merged.

### CODE REVIEW

File: fsharp-backend/src/LibExecution/StdLib/LibList.fs [hunk 1]
```diff
@@ -66,29 +66,24 @@ let fns : List<BuiltInFn> =
...
-//     ; sqlSpec = NotYetImplementedTODO
-//       ; previewable = Pure
-//     ; deprecated = NotDeprecated }
+    { name = fn "List" "tail" 0
+      parameters = [Param.make "list" (TList varA) ""]
+      returnType = TList varA
+      description =
+        "If the list contains at least one value, returns `Just` a list of every value other than the first. Otherwise, returns `Nothing`."
+      fn =
+        (* This matches Elm's implementation, with the added benefit that the error rail
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/StdLib/LibList.fs:70]: Can you change these comments to use `//` instead of `(*`?

File: fsharp-backend/src/LibExecution/StdLib/LibList.fs [hunk 2]
```diff
@@ -45,59 +45,51 @@ let fns : List<BuiltInFn> =
...
+         * means you don't need to handle unwrapping the option
+         * unless the passed list is truly empty (which shouldn't happen in most practical uses). *)
+        (function
+        | _, [ DList l ] ->
+            (match List.tryLast l with
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/StdLib/LibList.fs:75]: TryLast is very slow, it recurses to the end of the list. I think you could use `if List.length l > 0` instead.

File: fsharp-backend/tests/testfiles/list.tests [hunk 3]
```diff
@@ -73,3 +73,9 @@ List.filter_v0 [-20; 5; 9;] (fun x -> x < 1) = [-20]
 List.all_v0 [0; 1; 2; 3;] (fun elem -> elem > 0) = false
 List.all_v0 [1; 2; 3;] (fun elem -> elem > 0) = true
 List.all_v0 [0; -1; -2] (fun elem -> elem < 0) = false
+List.empty_v0 = []
+List.tail_v0 [10, 20, 30, 40] = [20, 30, 40]
+List.tail_v0 [] = Nothing
+List.head_v2 [1, 2, 3] = 1
+List.head_v2 [] = Nothing
+List.last_v0 [1000, 2000, 3000] =  3000
```
Comment (pbiggar) [inline fsharp-backend/tests/testfiles/list.tests:81]: Can you add a test for last on the empty list?

Can you add a List.last_v1 that behaves like List.head_v2?

**Review by pbiggar — CHANGES_REQUESTED** [review]
Nice job. A couple small changes and we're good to go.

File: fsharp-backend/src/LibExecution/StdLib/LibList.fs [hunk 4]
```diff
@@ -118,29 +118,26 @@ let fns : List<BuiltInFn> =
...
+        | _, [ DList l ] ->
+            if l.Length > 0 then List.last l else DNull
+            |> Value
+        | args -> incorrectArgs ())
+      sqlSpec = NotYetImplementedTODO
+      previewable = Pure
+      deprecated = ReplacedBy(fn "" "" 0) }
+    { name = fn "List" "last" 1
+      parameters = [Param.make "list" (TList varA) ""]
+      returnType = varA
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/StdLib/LibList.fs:127]: `varA` should be `TOption varA`

File: fsharp-backend/src/LibExecution/StdLib/LibList.fs [hunk 5]
```diff
@@ -118,29 +118,26 @@ let fns : List<BuiltInFn> =
...
+      description = "Returns the last value in `list`, wrapped in an option (`Nothing` if the list is empty)."
+      fn =
+        (function
+        | _, [DList l] ->
+            if l.Length > 0 then List.last l else DOption None
```
Comment (pbiggar) [inline fsharp-backend/src/LibExecution/StdLib/LibList.fs]: I think this fn can just be `Value (DOption (List.tryLast l))`, right?

File: fsharp-backend/tests/testfiles/list.tests [hunk 6]
```diff
@@ -79,4 +79,6 @@ List.tail_v0 [] = Nothing
 List.head_v2 [1, 2, 3] = 1
 List.head_v2 [] = Nothing
 List.last_v0 [1000, 2000, 3000] =  3000
-List.last_v0 [] =  null
+List.last_v0 [] = null
+List.last_v1 [1, 2, 3] =  3
```
Comment (pbiggar) [inline fsharp-backend/tests/testfiles/list.tests:83]: Should be `Some 3`, no?

**Review by pbiggar — CHANGES_REQUESTED** [review]
Looks nice, thanks! A few more small issues and we can merge.

File: fsharp-backend/src/LibExecution/StdLib/LibList.fs (hunk 5 above)
Comment (Athinanarof) [reply fsharp-backend/src/LibExecution/StdLib/LibList.fs]: Right!

File: fsharp-backend/tests/testfiles/list.tests (hunk 6 above)
Comment (Athinanarof) [reply fsharp-backend/tests/testfiles/list.tests]: It seems it is Just 3.

Comment (pbiggar) [reply fsharp-backend/tests/testfiles/list.tests]: Perfect.
